qe/drivers.py

```diff
@@ -1046,85 +363,130 @@ def do_diagonal(det_indices, psi_i, det_to_index_j, phase):
                 pass
 
     @staticmethod
-    def do_single(det_indices_i, phasemod, occ, h, p, psi_i, det_to_index_j, spin, exci):
-        # Single excitation from h to p, occ is index of orbital occupied
-        # Excitation is from internal to external space
-        # TODO: Some sort of additional pre-filtering based on constraints?
-        for a in det_indices_i:  # Loop through candidate determinants in internal space
-            det = psi_i[a]
-            excited_spindet = exci.apply_excitation(getattr(det, spin), [[h], [p]])
+    def do_single(
+        det_indices: List[int],
+        phasemod: int,
+        occ: OrbitalIdx,
+        h: OrbitalIdx,
+        p: OrbitalIdx,
+        psi_internal: Psi_det,
+        det_to_index: Dict[Determinant, int],
+        spin: str,
+    ):
+        """Do all single excitations from hole h -> particle p; occ is index of orbital that is necessarily occupied
+        Excitations from pre-filtered list of candidate |Determinant|s indicated by `det_indices`
+        Called by category functions corresponding to single excitations
+
+        For use in building the Hamiltonian in the variational step"""
+        for I in det_indices:
+            det = psi_internal[I]  # Grab |Determinant| I from w.f.
             if spin == "alpha":
-                excited_det = Determinant(excited_spindet, getattr(det, "beta"))
-            else:
-                excited_det = Determinant(getattr(det, "alpha"), excited_spindet)
-            if excited_det in det_to_index_j:  # Check if excited det is in external space
-                phase = phasemod * PhaseIdx.single_phase(getattr(det, spin), excited_spindet, h, p)
-                yield (a, det_to_index_j[excited_det]), phase
+                # Create new |Determinant| via excitation from h -> p
+                # Hole, particle are alpha spin orbitals
+                excited_det = det.apply_excitation(((h,), (p,)), ((), ()))
```

Comment:
Maybe better to have 2 type of excitation? Kind of July to pass lot of empty tuple

Reply:
Yeah we could do this. An `apply_single_exc`, `apply_double_samespin_exc`, and `apply_double_oppspin_exc`